### PR TITLE
fix tls handshake/error handling

### LINKED ISSUE #173: SSLError "no shared cipher"
Hi!

I noticed that some clients, when connecting to SMTP with STARTTLS, don't have a cipher that matches the one enabled on my end. This causes a SSLError with the message "No shared cipher"

```
SSLError: [SSL: NO_SHARED_CIPHER] no shared cipher (_ssl.c:1056)
  File "asyncio/sslproto.py", line 625, in _on_handshake_complete
    raise handshake_exc
  File "asyncio/sslproto.py", line 189, in feed_ssldata
    self._sslobj.do_handshake()
  File "ssl.py", line 763, in do_handshake
    self._sslobj.do_handshake()
```

This exception is logged via asyncio and is caused in the SSL wrapper.
I'm wondering if there is a way to catch this issue?

### PR DESCRIPTION
**Pull request by Elias481**
## What do these changes do?

Implement waiting for SSL setup/handshake within STARTTLS handler to be able to catch and handle (log) errors and to avoid session hanging around until timeout in such cases.
Add sessions peer information to some logging output where it was missing.
## Are there changes in behavior for the user?

Better logging of TLS handshake issues and less ressource consumption on such.
## Related issue number
Closes #173

## Checklist

- [X] I think the code is well written
- [X] Unit tests for the changes exist
- [X] Documentation reflects the changes
- [ ] Add a new news fragment into the `CHANGES` folder
  * name it `<issue_id>.<type>` (e.g. `588.bugfix`)
  * if you don't have an `issue_id` change it to the pr id after creating the PR
  * ensure type is one of the following:
    * `.feature`: Signifying a new feature.
    * `.bugfix`: Signifying a bug fix.
    * `.doc`: Signifying a documentation improvement.
    * `.removal`: Signifying a deprecation or removal of public API.
    * `.misc`: A ticket has been closed, but it is not of interest to users.
  * Make sure to use full sentences with correct case and punctuation, for example: `Fix issue with non-ascii contents in doctest text files.`

### CODE REVIEW

**Comment by waynew** [comment]
Hey @Elias481 thanks for the PR! I will review this sometime this week - probably on Saturday.

**Comment by waynew** [comment]
Sorry about the delay - life got a busy, reviewing now!

File: aiosmtpd/smtp.py [hunk 1]
```diff
@@ -317,21 +321,27 @@ def _set_rset_state(self):
...
+                log.info('%r Connection lost during _handle_client()',
+                         self.session.peer)
                 self._writer.close()
                 raise
             except Exception as error:
                 try:
                     status = await self.handle_exception(error)
-                    await self.push(status)
-                except Exception as error:
+                except Exception as error2:
```
Comment (waynew) [inline aiosmtpd/smtp.py]: Would this be better as `inner_error`? `error2` just seems a bit ambiguous to me :thinking:

File: aiosmtpd/smtp.py [hunk 2]
```diff
@@ -317,21 +321,27 @@ def _set_rset_state(self):
...
+                        log.exception('%r Exception in handle_exception()',
+                                      self.session.peer)
                         status = '500 Error: ({}) {}'.format(
-                            error.__class__.__name__, str(error))
+                            error2.__class__.__name__, str(error2))
                     except Exception:
                         status = '500 Error: Cannot describe error'
-                    await self.push(status)
+                finally:
+                    if isinstance(error, TLSSetupException):
```
Comment (waynew) [inline aiosmtpd/smtp.py]: Forgive me if I've missed something - but I don't see `error` defined other than in the exception. An example might help:

```
>>> try:
...   print('hi')
... except Exception as error:
...   ...
... finally:
...   print('error?', error)
... 
hi
Traceback (most recent call last):
  File "<stdin>", line 6, in <module>
NameError: name 'error' is not defined
```

If the code is successful, won't this fail with `NameError`?

File: aiosmtpd/smtp.py (hunk 1 above)
Comment (Elias481) [reply aiosmtpd/smtp.py]: Of course no problem (I'm just used to call them ex, ex2 for my personal codes so are used to that).

**Comment by pepoluan** [comment]
Do we have progress on this, or is this PR abandoned?

**Comment by pepoluan** [comment]
Hi @Elias481 , are you still up for this PR?

Or do you want to move this to [help-wanted] status?

File: aiosmtpd/smtp.py (hunk 2 above)
Comment (Elias481) [reply aiosmtpd/smtp.py:340]: I cannot see my comment anymore.  
I is defined in the main exception handler (line 328)

**Comment by Elias481** [comment]
Hi @pepoluan ,
I waited for progress on Your side, I'm not able to merge it.
So I can do a rebase to current version and have a look in the code-coverage issue reported for 3.7 if someone will merge it on time. If it will be pending afterwards again I can also leave it as it is now.

**Comment by pepoluan** [comment]
I see in Travis that there's a bunch of `asyncio:SSL handshake failed` errors there. But only for Python 3.7.

Maybe you can try merge/rebase from latest master; there's a whole LOT of changes that fixed things up.

**Comment by Elias481** [comment]
@pepoluan ok I can rebase/merge then.
The SSL errors are intended (it's a test for the exceptions to appear with the desired infromation) - they are only a bit more verbose in 3.7.
The issue that it is failing is test coverage at 99% only in one file for python 3.7. I can take a look on that after merging.

**Comment by Elias481** [comment]
So looks good now after merging, also with 3.9 and pypy3 tests no issues and merging was a no-brainer (thought did not verify the automtic merges besides test results are fine).

**Comment by pepoluan** [comment]
> 
> 
> So looks good now after merging, also with 3.9 and pypy3 tests no issues and merging was a no-brainer (thought did not verify the automtic merges besides test results are fine).

Awesome! Hopefully I can find some time tonight for testing your changes.

**Comment by pepoluan** [comment]
Uhhh... I totally forgot about this one.

I'll pick this up later today.

**Comment by pepoluan** [comment]
Fixed some test mismatch (451 --> 500), and now undergoing testing.

- [ ] Windows 10 (via PyCharm's runner): `(ALL)`

**This post is undergoing edits!**

**Review by pepoluan — APPROVED** [review]
I have no objections to the code & the tests. 🚀

**Pull request merged**